src/k2/tso/service/Service.cpp

```diff
@@ -143,12 +79,94 @@ seastar::future<> TSOService::_collectWorkerURLs() {
+                        cpu_set_t cpuset;
+                        CPU_ZERO(&cpuset);
+                        CPU_SET(pinCPU, &cpuset);
+                        if (0 != pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset)) {
+                            throw std::runtime_error("Unable to set affinity");
+                        }
+                    }
+                    while (_keepRunningPoller.test_and_set()) {
+                        GPSClockInst.poll();
+                    }
+                });
+                startFut = startFut.then([] {
+                    // notify all workers that the clock is initialized and we can start
+                    return k2::AppBase().getDist<TSOService>().invoke_on_all([] (auto& svc) {
+                        svc._clockInitialized.set_value();
+                    });
+                });
+            }
+            return startFut;
+        })
+        .then([this] {
+            K2LOG_I(log::tsoserver, "TSOService waiting for clock initialization");
+            // wait for clock to be initialized
+            return _clockInitialized.get_future();
+        })
+        .then([this, tsoID] {
+            K2LOG_I(log::tsoserver, "TSOService assigning tsoID");
+            _tsoId = tsoID;
+            return _collectWorkerURLs();
+        })
+        .then([this] {
+            // register RPC APIs
+            K2LOG_I(log::tsoserver, "TSOService registering RPC");
+            RPC().registerRPCObserver<dto::GetServiceNodeURLsRequest, dto::GetServiceNodeURLsResponse>
+            (dto::Verbs::GET_TSO_SERVICE_NODE_URLS, [this](dto::GetServiceNodeURLsRequest&& request) {
+                return _handleGetServiceNodeURLs(std::move(request));
+            });
+
+            RPC().registerRPCObserver<dto::GetTimestampRequest, dto::GetTimestampResponse>(dto::Verbs::GET_TSO_TIMESTAMP, [this](dto::GetTimestampRequest&& request) {
+                return _handleGetTimestamp(std::move(request));
+            });
+
+            K2LOG_I(log::tsoserver, "TSOService started with ID {}, error bound {}", _tsoId, _CPOErrorBound);
+            return seastar::make_ready_future<>();
+        });
+}
 
-    K2LOG_I(log::tsoserver, "Registration with CPO successful with id {}", _tsoId);
-    return seastar::make_ready_future<>();
+seastar::future<std::tuple<Status, dto::AssignTSOResponse>>
+TSOService::_handleAssignment(dto::AssignTSORequest&& request) {
+    if (request.tsoErrBound <= 0s) {
+        return RPCResponse(Statuses::S400_Bad_Request("TSO error bound cannot be <= 0s"), dto::AssignTSOResponse{});
+    }
+    return _assign(request.tsoID, request.tsoErrBound).then([] {
+        return RPCResponse(Statuses::S200_OK("OK"), dto::AssignTSOResponse{});
+    }).handle_exception([] (auto exp) {
+        K2LOG_W_EXC(log::tsoserver, exp, "failed to assign");
+        return RPCResponse(Statuses::S500_Internal_Server_Error("Unable to assign TSO server"), dto::AssignTSOResponse{});

```

Comment:
It would be helpful to add the message from the exception here as part of the response